docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,474 @@
+      * **verification**: section for providing TLS verification details.
+        * **skip**: defines if proxy should skip SAN verification. Default `false`.
+        * **subjectAltNames**: list of names to verify in the certificate.
+      * **caCert**: defines a certificate of CA.
+        * one of `inline`, `inlineString` or `secret`.
+      * **clientCert**: defines a certificate of a client.
+        * one of `inline`, `inlineString` or `secret`.
+      * **clientKey**: defines a client private key.
+        * one of `inline`, `inlineString` or `secret`.
+* **status**: status of an object managed by Kuma control-plane
+  * **vips**: list of allocated VIPs
+    * **ip**: allocated IP for a provided domain with `InternalVIP` type in a match section
+    * **type**: provides information about the way IP was provided
+    * **hostname**: provides a domain with `InternalVIP` type in a match section for which IP was allocated. In case of many entries it helps corelating entries.
+
+#### Cluster name
+
+Currently, we name clusters after their respective policies. However, when dealing with multiple protocols, we find it necessary to include the port number in the cluster name.
+Example:
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: httpbin
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: Domain
+    value: httpbin.com
+    port: 80
+    protocol: http2
+  - type: Domain
+    value: httpbin.com
+    port: 443
+    protocol: tls
+ destination:
+   type: Passthrough
+```
+
+In this case, for port `80`, we require HTTP/2 configuration in the cluster settings. This necessitates having a separate configuration for each cluster.
+
+clusterName: `{policyName}_{port}`, so for the above policy name would looks like: `httpbin_80` and `httpbin_443`
+
+#### Extensability
+
+Provided model allows creating separate plugins which can integrate with different cloud providers or components. The user can create it's own custom configuration and later in the code create XDS configuration based on it. 
+
+Example:
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP
+    value: example.ext.svc.local
+    port: 80
+    protocol: http
+ destination:
+   type: Extension
+   extension:
+     type: Lambda 
+     config:
+       arn: arn:aws:lambda:us-west-2:123456789012:function:my-function
+```
+
+#### IP intersections 
+
+It's possible for users to create a policy where the IP ranges intersect with each other. We cannot validate this during policy creation, but during configuration generation, we are able to check if IPs intersect. In such cases, we can log a message stating:
+
+> external-service1 and external-service2 have overlapping IPs X.X.X.X, which can disrupt your traffic.
```

Comment:
We can consider conditions in status and controller to fill it out, but I'm not sure it's worth it now. What happens in this scenario? Which external service takes precedence?

Reply:
> I'm not sure it's worth it now

agreed

> What happens in this scenario? Which external service takes precedence?

I don't think we should give any promises on the order just document it as undefined behaviour and pick the first one lexicographically? WDYT?

Reply:
technically "created first" is more safe, but we do "lexicographically" in case of conflicts in other places

Reply:
Added note